### ISSUE TITLE
CIDR allocation slow with IP aliases

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#52284]
Recently while starting a large cluster with IP aliases enabled, I observed that the CIDR allocation was too slow. This leads to cluster startup failing the validation step with some nodes unready.
There seem to be no issues on the GCE side as the nodes already have the alias IP range assigned. The issue is with cloud-CIDR-controller (range allocator is fine). It's slow because of at least 2 reasons:
- processing work queue serially (unlike range allocator which does it concurrently)
- blocking the main control loop on nodeSpec update (this should be done through a separate channel)

I have a fix for this in progress. Besides these, there are multiple other improvements that can be made to cidr allocators. I'll file a separate issue for it.

cc @kubernetes/sig-network-bugs @kubernetes/sig-scalability-misc @wojtek-t @bowei @thockin

**Comment by shyamjvs** [comment]
Adding 1.8 milestone as I think this is pretty important for large clusters. Feel free to change if you feel otherwise.

**Comment by shyamjvs** [comment]
Rate at which nodes are getting CIDRs assigned with and without this change (actually it's about nodes getting ready, but for large clusters it's currently bottle-necked by cidr-allocation):

```
Waiting for 1001 ready nodes. 2 ready nodes, 41 registered. Retrying.
Waiting for 1001 ready nodes. 19 ready nodes, 217 registered. Retrying.
Waiting for 1001 ready nodes. 31 ready nodes, 383 registered. Retrying.
Waiting for 1001 ready nodes. 42 ready nodes, 552 registered. Retrying.
Waiting for 1001 ready nodes. 62 ready nodes, 726 registered. Retrying.
Waiting for 1001 ready nodes. 79 ready nodes, 900 registered. Retrying.
Waiting for 1001 ready nodes. 95 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 116 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 132 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 151 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 161 ready nodes, 1001 registered. Retrying.
```

```
Waiting for 1001 ready nodes. 106 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 118 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 141 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 154 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 181 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 212 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 236 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 251 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 280 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 307 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 331 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 342 ready nodes, 1001 registered. Retrying.
Waiting for 1001 ready nodes. 376 ready nodes, 1001 registered. Retrying.
```

**Comment by wojtek-t** [comment]
@shyamjvs - how frequently this is putting this information - without this information is mostly useless.

BTW - it still seems pretty slow - I would expect the bottleneck for that to be QPS limits on the client side. If not, what is bottleneck there?

**Comment by shyamjvs** [comment]
It's once every 15s.

> BTW - it still seems pretty slow - I would expect the bottleneck for that to be QPS limits on the client side. If not, what is bottleneck there?

Yes, it's still slow. The bottleneck is not qps, but `get instance` calls to cloud-provider. I've explained it in greater detail here - https://github.com/kubernetes/kubernetes/pull/52285#issuecomment-328808398
We need to use lists instead of gets.

**Comment by spiffxp** [comment]
friendly v1.8 release ping, does this issue need to be in the v1.8 milestone? is this issue actively being worked?

**Comment by bowei** [comment]
@shyamjvs Your current patch seems to be ok -- we can do the broader fix for 1.9. What do you think?
